kernel/src/table_configuration.rs

```diff
@@ -176,29 +181,199 @@ impl TableConfiguration {
+            }
+            _ => Vec::new(),
+        }
+    }
+
+    /// Returns `Ok` if the kernel supports reading from this table. This checks that the
+    /// protocol's reader features are all supported.
+    fn ensure_read_supported(&self) -> DeltaResult<()> {
+        // Version check
+        match self.protocol.min_reader_version() {
+            1..=3 => {}
+            _ => {
+                return Err(Error::unsupported(format!(
+                    "Unsupported minimum reader version {}",
+                    self.protocol.min_reader_version()
+                )))
+            }
+        }
+
+        // Check all enabled reader features have kernel read support
+        for feature in self.get_enabled_reader_features() {
+            let Some(info) = feature.info() else {
+                return Err(Error::unsupported(format!("Unknown feature '{}'", feature)));
+            };
+
+            // Check read support
+            match &info.read_support {
+                KernelSupport::Supported => Ok(()),
+                KernelSupport::NotSupported => Err(Error::unsupported(format!(
+                    "Feature '{}' not supported for reads",
+                    info.name
+                ))),
+                KernelSupport::Custom(check) => {
+                    check(&self.protocol, &self.table_properties, Operation::Scan)
+                }
+            }?;
+
+            // Validate feature requirements
+            self.validate_feature_requirements(info.name, info.feature_requirements)?;
+        }
+
+        Ok(())
+    }
+
     /// Returns `true` if the kernel supports writing to this table. This checks that the
     /// protocol's writer features are all supported.
     #[internal_api]
     pub(crate) fn ensure_write_supported(&self) -> DeltaResult<()> {
-        self.protocol.ensure_write_supported()?;
+        // Version check
+        match self.protocol.min_writer_version() {
+            1 | 2 | 7 => {}
+            _ => {
+                return Err(Error::unsupported(
+                    "Currently delta-kernel-rs can only write to tables with protocol.minWriterVersion = 1, 2, or 7",
+                ))
+            }
+        }
+
+        // Check all enabled writer features have kernel write support
+        for feature in self.get_enabled_writer_features() {
+            let Some(info) = feature.info() else {
+                return Err(Error::unsupported(format!("Unknown feature '{}'", feature)));
+            };
+
+            // Check write support
+            match &info.write_support {
+                KernelSupport::Supported => Ok(()),
+                KernelSupport::NotSupported => Err(Error::unsupported(format!(
+                    "Feature '{}' not supported for writes",
+                    info.name
+                ))),
+                KernelSupport::Custom(check) => {
+                    check(&self.protocol, &self.table_properties, Operation::Scan)
+                }
+            }?;
```

Comment:
let's get rid of the sneaky `?` and return errors when needed.
```suggestion
            match &info.write_support {
                KernelSupport::Supported => {},
                KernelSupport::NotSupported => return Err(Error::unsupported(format!(
                    "Feature '{}' not supported for writes",
                    info.name
                ))),
                KernelSupport::Custom(check) => {
                    check(&self.protocol, &self.table_properties, Operation::Scan)?
                }
            };
```